File: floating_palette/windows/services/host_service.cpp
#include "host_service.h"

#include "../core/logger.h"

namespace floating_palette {

void HostService::Handle(
    const std::string& command,
    const std::string* window_id,
    const flutter::EncodableMap& params,
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  if (command == "getProtocolVersion") {
    GetProtocolVersion(std::move(result));
  } else if (command == "getCapabilities") {
    GetCapabilities(std::move(result));
  } else if (command == "getServiceVersion") {
    GetServiceVersion(params, std::move(result));
  } else if (command == "getSnapshot") {
    GetSnapshot(std::move(result));
  } else if (command == "ping") {
    Ping(std::move(result));
  } else {
    result->Error("UNKNOWN_COMMAND", "Unknown host command: " + command);
  }
}

void HostService::GetProtocolVersion(
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  result->Success(flutter::EncodableValue(flutter::EncodableMap{
      {flutter::EncodableValue("version"),
       flutter::EncodableValue(kProtocolVersion)},
      {flutter::EncodableValue("minDartVersion"),
       flutter::EncodableValue(kMinDartVersion)},
      {flutter::EncodableValue("maxDartVersion"),
       flutter::EncodableValue(kMaxDartVersion)},
  }));
}

void HostService::GetCapabilities(
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  result->Success(flutter::EncodableValue(flutter::EncodableMap{
      {flutter::EncodableValue("blur"), flutter::EncodableValue(true)},
      {flutter::EncodableValue("transform"), flutter::EncodableValue(false)},
      {flutter::EncodableValue("globalHotkeys"),
       flutter::EncodableValue(true)},
      {flutter::EncodableValue("glassEffect"),
       flutter::EncodableValue(false)},
      {flutter::EncodableValue("multiMonitor"),
       flutter::EncodableValue(true)},
      {flutter::EncodableValue("contentSizing"),
       flutter::EncodableValue(true)},
      {flutter::EncodableValue("platform"),
       flutter::EncodableValue("windows")},
  }));
}

void HostService::GetServiceVersion(
    const flutter::EncodableMap& params,
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  auto it = params.find(flutter::EncodableValue("service"));
  if (it == params.end() || !std::holds_alternative<std::string>(it->second)) {
    result->Error("INVALID_PARAMS", "Missing 'service' parameter");
    return;
  }
  const auto& service = std::get<std::string>(it->second);
  result->Success(flutter::EncodableValue(flutter::EncodableMap{
      {flutter::EncodableValue("service"),
       flutter::EncodableValue(service)},
      {flutter::EncodableValue("version"), flutter::EncodableValue(1)},
  }));
}

void HostService::GetSnapshot(
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  auto all = WindowStore::Instance().All();
  flutter::EncodableList snapshot;

  for (auto& [id, window] : all) {
    if (!window || !window->hwnd || window->is_destroyed) continue;

    RECT rect;
    bool has_rect = GetWindowRect(window->hwnd, &rect) != FALSE;

    bool visible = ::IsWindowVisible(window->hwnd) != FALSE;
    bool focused = (GetForegroundWindow() == window->hwnd);

    flutter::EncodableMap entry{
        {flutter::EncodableValue("id"), flutter::EncodableValue(id)},
        {flutter::EncodableValue("visible"), flutter::EncodableValue(visible)},
        {flutter::EncodableValue("x"),
         flutter::EncodableValue(
             has_rect ? static_cast<double>(rect.left) : 0.0)},
        {flutter::EncodableValue("y"),
         flutter::EncodableValue(
             has_rect ? static_cast<double>(rect.top) : 0.0)},
        {flutter::EncodableValue("width"),
         flutter::EncodableValue(
             has_rect ? static_cast<double>(rect.right - rect.left) : 0.0)},
        {flutter::EncodableValue("height"),
         flutter::EncodableValue(
             has_rect ? static_cast<double>(rect.bottom - rect.top) : 0.0)},
        {flutter::EncodableValue("focused"),
         flutter::EncodableValue(focused)},
    };
    snapshot.push_back(flutter::EncodableValue(entry));
  }

  result->Success(flutter::EncodableValue(snapshot));
}

void HostService::Ping(
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  result->Success(flutter::EncodableValue(true));
}

}  // namespace floating_palette

File: floating_palette/macos/Classes/services/HostService.swift
import Cocoa
import FlutterMacOS

/// Service for host-level operations.
///
/// Handles protocol version, capabilities, and window snapshots.
class HostService {
    /// Protocol version - increment when API changes.
    private static let protocolVersion = 1

    /// Minimum Dart version this native plugin supports.
    private static let minDartVersion = 1

    /// Maximum Dart version this native plugin supports.
    private static let maxDartVersion = 1

    private var eventSink: ((String, String, String?, [String: Any]) -> Void)?

    func setEventSink(_ sink: @escaping (String, String, String?, [String: Any]) -> Void) {
        self.eventSink = sink
    }

    func handle(_ command: String, windowId: String?, params: [String: Any], result: @escaping FlutterResult) {
        switch command {
        case "getProtocolVersion":
            result([
                "version": HostService.protocolVersion,
                "minDartVersion": HostService.minDartVersion,
                "maxDartVersion": HostService.maxDartVersion,
            ])

        case "getCapabilities":
            result(getCapabilities())

        case "getServiceVersion":
            if let service = params["service"] as? String {
                result(getServiceVersion(service))
            } else {
                result(FlutterError(code: "INVALID_PARAMS", message: "Missing 'service' parameter", details: nil))
            }

        case "getSnapshot":
            result(getSnapshot())

        case "ping":
            result(true)

        default:
            result(FlutterError(code: "UNKNOWN_COMMAND", message: "Unknown host command: \(command)", details: nil))
        }
    }

    // MARK: - Capabilities

    private func getCapabilities() -> [String: Any] {
        let osVersion = ProcessInfo.processInfo.operatingSystemVersionString

        return [
            "blur": true,
            "transform": true,
            "globalHotkeys": true,
            "glassEffect": true,
            "multiMonitor": true,
            "contentSizing": true,
            "textSelection": true,
            "platform": "macos",
            "osVersion": osVersion,
        ]
    }

    // MARK: - Service Versions

    private func getServiceVersion(_ service: String) -> [String: Any] {
        // All services are at version 1
        let versions: [String: Int] = [
            "window": 1,
            "visibility": 1,
            "frame": 1,
            "transform": 1,
            "animation": 1,
            "input": 1,
            "focus": 1,
            "zorder": 1,
            "appearance": 1,
            "screen": 1,
            "message": 1,
            "host": 1,
        ]

        return [
            "service": service,
            "version": versions[service] ?? 0,
        ]
    }

    // MARK: - Snapshot (Hot Restart Recovery)

    private func getSnapshot() -> [[String: Any]] {
        var snapshots: [[String: Any]] = []

        for (id, window) in WindowStore.shared.all() {
            let frame = window.frame
            let isVisible = window.isVisible
            let isKey = window.panel.isKeyWindow

            snapshots.append([
                "id": id,
                "visible": isVisible,
                "x": frame.origin.x,
                "y": frame.origin.y,
                "width": frame.size.width,
                "height": frame.size.height,
                "focused": isKey,
            ])
        }

        return snapshots
    }

    // MARK: - Events

    func sendAppActivated() {
        eventSink?("host", "appActivated", nil, [:])
    }

    func sendAppDeactivated() {
        eventSink?("host", "appDeactivated", nil, [:])
    }

    func sendAppWillTerminate() {
        eventSink?("host", "appWillTerminate", nil, [:])
    }
}
